text/0000-cargo-plugin-fields.md

```diff
@@ -0,0 +1,119 @@
+- Feature Name: (fill me in with a unique ident, `cargo_plugin_fields`)
+- Start Date: (fill me in with today's date, 2019-10-03)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/cargo#7423](https://github.com/rust-lang/cargo/issues/7423)
+
+# Summary
+[summary]: #summary
+
+Allow Cargo.toml to contain arbitrary fields in place, which `cargo` itself is not aware of, and are given semantic meaning by plugins.
+
+# Motivation
+[motivation]: #motivation
+
+Plugins sometimes need additional information beyond that specified in Cargo.toml. Existing information like package.metadata is relegated to a side table.  The motivation for this equally, the ability to specify these plugin fields, and specify them _in place_.
```

Comment:
TOML is flexible in how it lets you express things. All of the following are equivalent:

```toml
package.metadata.pluginname.key = "value"
```

```toml
[package]
metadata.pluginname.key = "value"
```

```toml
[package.metadata]
pluginname.key = "value"
```

```toml
[package.metadata.pluginname]
key = "value"
```

```toml
[package]
metadata.pluginname = { key = "value" }
```

I’m not feeling this motivation.

A more interesting concept (which would be a completely different RFC) would be putting this in `dependencies.pluginname.metadata` and directly exposing it to the plugin somehow, so that you could write something like this in Cargo.toml:

```toml
[dependencies.pluginname]
version = "1.2.3"
metadata.key = "value"
```

Reply:
I think, I perhaps need to refocus this RFC to focus more upon using json-ld,
Toml is defined in such a way that every Toml document can be converted to a JSON document.

And refocus this on providing a translation layer from Cargo.toml to JSON-LD documents.
Where the keys them become "@pluginname:key", and the semantic differences between having an arbitrary values for fields.

For example in package.metadata, you can have 
```[package.metadata]
pluginname.foo = "0.0.1"
```

Where foo means nothing to `pluginname`, it expands it to package.metadata.pluginname = {name = "foo", version = "0.0.1"}

Ironically we can then give all cargo fields semantic meaning like "@cargo:package" as a translation to `[package]` to highlight the semantic difference, the one key that cargo key I know of to which we cannot give semantic meaning to is it's `[package.metadata]` which is capable of being an arbitrary table.  The only thing in a json-ld document we can treat it as is a json-literal.  Where the field 
```"@pluginname:name" = "foo"
"@pluginname:version" = "bar"
```
Would translate to json-ld triples encoding fields cargo doesn't have any prior knowledge of, but can be checked for validity none the less.

I'm not sure if it's best to recycle this based upon that, or start out with a new RFC proposal.